sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStageExec.scala

```diff
@@ -51,13 +51,30 @@ abstract class QueryStageExec extends LeafExecNode {
    */
   val plan: SparkPlan
 
+  /**
+   * Name of this query stage which is unique in the entire query plan.
+   */
+  val name: String = s"${this.getClass.getSimpleName}-$id"
+
+  /**
+   * This flag aims to detect if the stage materialization is started. This helps
+   * to avoid unnecessary stage materialization when the stage is canceled.
+   */
+  private val materializationStarted = new AtomicBoolean()
```

Comment:
sorry for the last-minute proposal, but I'm wondering if it's more efficient to push this cancelation optimization into shuffle and broadcast nodes.

It looks a bit fragile to operate on the `shuffleFuture` directly in `ShuffleQueryStageExec.cancel`. I think we should let `ShuffleExchangeLike` provide clear APIs to do it. Today it provides `submitShuffleJob`, and it should also provide `cancelShuffleJob`.

Within `ShuffleExchangeLike`, we can do more optimizations. e.g. even if we cancel the shuffle stage after the shuffle stage is submitted, we can still avoid submitting the shuffle job, as the shuffle node might be doing other preparation work: generating the shuffle dependency, waiting for subqueries to finish, etc. It's more efficient to check the isCanceled flag at the last minute, right before submitting the shuffle job.

Reply:
I agree on this and we need clean API definition so please find new refactoring on this direction by last commit.